File: day_1/locations.js
let fs = require("node:fs");

input = fs.readFileSync("./input.txt", "utf-8").split("\n");

let inputSplit = input.map((e) => e.split("   "));

let left = [];
let right = [];

inputSplit.forEach((e) => {
  if (e[0] === "") {
    return;
  } else {
    left.push(parseInt(e[0]));
    right.push(parseInt(e[1]));
  }
});

left.map((e) => {
  if (e === NaN) console.log(e);
});

let sorted_left = left.sort();
let sorted_right = right.sort();
let total = 0;
let index = 0;

while (index < sorted_left.length) {
  total += Math.abs(sorted_left[index] - sorted_right[index]);
  console.log(total);
  index++;
}

console.log(total);

File: day_1/locations.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

def separate_input
  sides = { left: [], right: [] }
  File.foreach('input.txt') do |line|
    pieces = line.split('   ')
    pieces[1] = pieces[1].chomp

    sides[:left] << pieces[0].to_i
    sides[:right] << pieces[1].to_i
  end
  sides
end

input = separate_input

# part one
sorted_left = input[:left].sort
sorted_right = input[:right].sort
index = 0
totals = []

until index == sorted_left.length
  difference = (sorted_left[index] - sorted_right[index]).abs
  totals << difference
  index += 1
end

p totals.sum

# part two

def create_hash(input)
  occurrences = {}
  input[:right].each do |num|
    if occurrences[num].nil?
      occurrences[num] = 1
    else
      occurrences[num] += 1
    end
  end
  occurrences
end

sim_right = create_hash(input)

def calculate_sim_score(right, input)
  sim_score = 0
  input[:left].each do |num|
    if right.include?(num) && right[num] > 1
      sim_score += (num * right[num])
    elsif right[num] == 1
      sim_score += num
    end
  end
  sim_score
end

p calculate_sim_score(sim_right, input)
